### PR TITLE
8297778: Modernize and improve module jdk.sctp

### PR DESCRIPTION
**Pull request by minborg**
This PR proposes a variety of modernisations to the `jdk.sctp` module.

During the fix of https://bugs.openjdk.org/browse/JDK-8296024, several improvement areas were identified including: 

* Replacing duplicate code segments 
* Making certain fields final 
* Using enhanced switch 
* Using records 
* Fixing typos 
* Marking fields participating in serialisation with `@Serial` 
* Modernizing toString() implementations 
* Using pattern matching 
* Using diamond operators

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8297778](https://bugs.openjdk.org/browse/JDK-8297778): Modernize and improve module jdk.sctp

### Reviewers
 * [Daniel Fuchs](https://openjdk.org/census#dfuchs) (@dfuch - **Reviewer**)
 * [Sergey Tsypanov](https://openjdk.org/census#stsypanov) (@stsypanov - Author) ⚠️ Review applies to [91b10aa9](https://git.openjdk.org/jdk/pull/11418/files/91b10aa935ac511ae1617a764cf2d7fa3b4e0d77)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11418/head:pull/11418` \
`$ git checkout pull/11418`

Update a local copy of the PR: \
`$ git checkout pull/11418` \
`$ git pull https://git.openjdk.org/jdk pull/11418/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11418`

View PR using the GUI difftool: \
`$ git pr show -t 11418`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11418.diff">https://git.openjdk.org/jdk/pull/11418.diff</a>

</details>

### CODE REVIEW

File: src/jdk.sctp/share/classes/sun/nio/ch/sctp/SctpStdSocketOption.java
Comment (dfuch) [inline src/jdk.sctp/share/classes/sun/nio/ch/sctp/SctpStdSocketOption.java:32]: This changes the implementation of `toString`, `equals` and `hashCode`. Have you verified that it is OK to do so?

File: src/jdk.sctp/unix/classes/sun/nio/ch/sctp/SctpMultiChannelImpl.java
Comment (dfuch) [inline src/jdk.sctp/unix/classes/sun/nio/ch/sctp/SctpMultiChannelImpl.java:33]: It would be better to avoid wildcards in regular imports.

**Review by dfuch — COMMENTED** [review]
It is a bit unusual to use a local class as a holder class (the typical manner is to a have a private static final nested class) but we didn't have local classes until recently - so maybe that's OK. I assume there's only one version of that class and it is static?

**Comment by AlanBateman** [comment]
The SCTP hasn't had much TLC in some time, maybe JDK 6, so not surprising there is a lot of cleanup/modernization.

You'll need to add a copyright header to sun.nio.ch. sctp.Util. It's a pity there are duplicate copies.

File: src/jdk.sctp/unix/classes/sun/nio/ch/sctp/SctpMultiChannelImpl.java
Comment (RogerRiggs) [inline src/jdk.sctp/unix/classes/sun/nio/ch/sctp/SctpMultiChannelImpl.java:722]: There should be a 1st sentence describing the function; its odd to see only an \@throws.

File: src/jdk.sctp/aix/classes/sun/nio/ch/sctp/Util.java
Comment (RogerRiggs) [inline src/jdk.sctp/aix/classes/sun/nio/ch/sctp/Util.java:1]: Can this be shared in **share**/classes/sun/nio/sctp/Util.java?

File: src/jdk.sctp/macosx/classes/sun/nio/ch/sctp/SctpChannelImpl.java
Comment (RogerRiggs) [inline src/jdk.sctp/macosx/classes/sun/nio/ch/sctp/SctpChannelImpl.java:44]: Going a bit beyond just updating syntax; but that's a different PR...
There could be more sharing of the unsupported implementation if SctpChannel was not abstract and its method bodies threw the exception.  There would less duplication for unsupported platforms.

Comment (AlanBateman) [reply src/jdk.sctp/macosx/classes/sun/nio/ch/sctp/SctpChannelImpl.java:44]: > There could be more sharing of the unsupported implementation if SctpChannel was not abstract and its method bodies threw the exception. There would less duplication for unsupported platforms.

I think this is accident of history. As I recall, the "unsupported implementation" was Windows only in JDK 6 but it has since been copied. It could be cleaned up.

**Review by stuart-marks — COMMENTED** [review]
Various changes such as fixing spelling errors, addition of `@Serial`, removing redundant type arguments, using `instanceof` patterns, and using switch expressions seem fine to me. However, I'll defer to the judgment of the maintainers of this code.

If a larger-scale refactoring is done to reduce the redundancy of UOE-throwing for the platforms where SCTP isn't supported, I'd recommend doing it in a separate bug/PR.

**Comment by minborg** [comment]
> It is a bit unusual to use a local class as a holder class (the typical manner is to a have a private static final nested class) but we didn't have local classes until recently - so maybe that's OK. I assume there's only one version of that class and it is static?

As I understand it, there are zero instances of the class as it is only declared. The class only contains a class variable. In a recent Java version, classes that are "effectively static" (my wording) (i.e. does not refernece the outer class) does not hold a reference to the outer class. So, it is not static but in this case, I believe it does not matter because 1) it is never instantiated and 2) it is "effectively static".

File: src/jdk.sctp/unix/classes/sun/nio/ch/sctp/SctpMultiChannelImpl.java
Comment (minborg) [reply src/jdk.sctp/unix/classes/sun/nio/ch/sctp/SctpMultiChannelImpl.java:33]: Agreed. I would have to find the settings in my IDE to prevent this automatic "feature".

File: src/jdk.sctp/macosx/classes/sun/nio/ch/sctp/SctpChannelImpl.java
Comment (minborg) [reply src/jdk.sctp/macosx/classes/sun/nio/ch/sctp/SctpChannelImpl.java:44]: Created https://bugs.openjdk.org/browse/JDK-8297822?filter=-2 and will move changes for this to a separate PR.

File: src/jdk.sctp/unix/classes/sun/nio/ch/sctp/SctpChannelImpl.java
Comment (stsypanov) [inline src/jdk.sctp/unix/classes/sun/nio/ch/sctp/SctpChannelImpl.java:1084]: Is it possible to use `import java.security.PrivilegedAction;` instead of fully-qualified name?

File: src/jdk.sctp/share/classes/com/sun/nio/sctp/SctpStandardSocketOptions.java
Comment (stsypanov) [inline src/jdk.sctp/share/classes/com/sun/nio/sctp/SctpStandardSocketOptions.java:53]: I guess you don't need fully qualified import for `sun.nio.ch.sctp.SctpStdSocketOption` as it is already imported in this class

File: src/jdk.sctp/unix/classes/sun/nio/ch/sctp/SctpChannelImpl.java
Comment (stsypanov) [inline src/jdk.sctp/unix/classes/sun/nio/ch/sctp/SctpChannelImpl.java:1085]: I think we can import `java.security.PrivilegedAction` instead of using its full name

File: src/jdk.sctp/unix/classes/sun/nio/ch/sctp/ResultContainer.java
Comment (minborg) [reply src/jdk.sctp/unix/classes/sun/nio/ch/sctp/ResultContainer.java:124]: That's what I also thought at first. I double checked in my IDE and there, it looks good. Is this a bug in GitHub's rendering?


image

also

```
pminborg@pminborg-mac minborg-jdk % cat /Users/pminborg/dev/minborg-jdk/open/src/jdk.sctp/unix/classes/sun/nio/ch/sctp/ResultContainer.java

...

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("Type: ");
        switch (type) {
            case NOTHING              -> sb.append("NOTHING");
            case MESSAGE              -> sb.append("MESSAGE");
            case SEND_FAILED          -> sb.append("SEND FAILED");
            case ASSOCIATION_CHANGED  -> sb.append("ASSOCIATION CHANGE");
            case PEER_ADDRESS_CHANGED -> sb.append("PEER ADDRESS CHANGE");
            case SHUTDOWN             -> sb.append("SHUTDOWN");
            default                   -> sb.append("Unknown result type");
        }
        sb.append(", Value: ");
        sb.append((value == null) ? "null" : value.toString());
        return sb.toString();
    }
}

```

**Comment by minborg** [comment]
`receiverThread` and `senderThread` are declared `volatile` but are always accessed (r/w) whilst synchronising on the `stateLock`. So, this indicates that the the fields do not need to be `volatile` and could instead be used with normal memory semantics?

**Comment by minborg** [comment]
> Btw, the same issue with volatile is present in `SctpMultiChannelImpl` and `SctpServerChannelImpl` as well, so you could have a look into them.

Fixed!

**Comment by stsypanov** [comment]
> receiverThread and senderThread are declared volatile but are always accessed (r/w) whilst synchronising on the stateLock.

I think it's better to keep them as is. What I'm saying is that you don't need explicit assignment of default value for a volatile field as it costs us, and without this assignment the default value will be there by the spec.

P.S. I guess you forgot about `private volatile long thread = 0;` in `SctpServerChannelImpl` and unnecessary fully-qualified names in `SctpStandardSocketOptions`

**Comment by minborg** [comment]
> > receiverThread and senderThread are declared volatile but are always accessed (r/w) whilst synchronising on the stateLock.
> 

@stsypanov Yes I am aware that this was a separate observation which is not related to the assignment of the variables. I can surely keep them volatile but I wonder why they are volatile in the first place.

**Comment by AlanBateman** [comment]
> > > receiverThread and senderThread are declared volatile but are always accessed (r/w) whilst synchronising on the stateLock.
> 
> @stsypanov Yes I am aware that this was a separate observation which is not related to the assignment of the variables. I can surely keep them volatile but I wonder why they are volatile in the first place.

The SCTP implementation dates from JDK 7 and has not changed much since then.  In the case of the thread IDs used for signalling then I suspect this was copied from the original SocketChannelImpl code that has been replaced/re-written by way of splitting the blocking and non-blocking code paths (the thread IDs are only needed for channels configured blocking). If all access to the thread IDs is guarded by stateLock then you should be okay to making it non-volatile. Testing is a challenge when touching this area as SCTP requires special kernel configuration. Keeping the changes simple and easy to review would be helpful.

**Review by dfuch — APPROVED** [review]
Overall this looks like a good cleanup. Please do not integrate before running all related tests + tier1 and tier2.

**Review by stsypanov — APPROVED** [review]

File: src/jdk.sctp/share/classes/com/sun/nio/sctp/SctpStandardSocketOptions.java
Comment (RogerRiggs) [inline src/jdk.sctp/share/classes/com/sun/nio/sctp/SctpStandardSocketOptions.java:72]: The indentation here looks to be missing.  (And was before).

File: src/jdk.sctp/share/classes/com/sun/nio/sctp/SctpStandardSocketOptions.java
Comment (RogerRiggs) [inline src/jdk.sctp/share/classes/com/sun/nio/sctp/SctpStandardSocketOptions.java:162]: Please fix the indentation, here and in the rest of the file/files.

**Review by dfuch — APPROVED** [review]

**Comment by dfuch** [comment]
I will sponsor this PR after the RDP 1 fork.

**Pull request closed without merging**
